pandas/core/arraylike.py

```diff
@@ -157,11 +157,67 @@ def array_ufunc(self, ufunc: Callable, method: str, *inputs: Any, **kwargs: Any)
     --------
     numpy.org/doc/stable/reference/arrays.classes.html#numpy.class.__array_ufunc__
     """
+    from pandas.core.frame import DataFrame
     from pandas.core.generic import NDFrame
     from pandas.core.internals import BlockManager
 
     cls = type(self)
 
+    is_ndframe = [isinstance(x, NDFrame) for x in inputs]
+    is_frame = [isinstance(x, DataFrame) for x in inputs]
+
+    if (sum(is_ndframe) >= 2) and (sum(is_frame) >= 1):
+        # if there are 2 alignable inputs, of which at least 1 is a
+        # DataFrame -> we would have had no alignment before -> warn that this
+        # will align in the future
+
+        # the first frame is what determines the output index/columns in pandas < 1.2
+        for x in inputs:
+            if isinstance(x, DataFrame):
+                first_frame = x
+                break
+
+        # check if the objects are aligned or not
+        def is_aligned(frame, other):
+            if isinstance(other, DataFrame):
+                return frame._indexed_same(other)
+            else:
+                # Series -> match index
+                return frame.columns.equals(other.index)
+
+        non_aligned = sum(
+            not is_aligned(first_frame, x) for x in inputs if isinstance(x, NDFrame)
+        )
+
+        # if at least one is not aligned -> warn and fallback to array behaviour
+        if non_aligned:
+            warnings.warn(
+                "Calling a ufunc on non-aligned DataFrames/Series. Currently, the "
```

Comment:
because the Series behavior is different, this warning could be misleading?

Reply:
Hmm, yes. The most explicit is "non-aligned DataFrames or DataFrame/Series combination" or something like that, but wanted to keep it shorter .. 
I agree the current can be misleading though (although you will of course never see the warning with only series)